README.md
_This is an archive of escripting.com, which I used to publish little bits of code to circa 2001 - 2005._

eScripting brings you high quality web-based tools to make your work easier. All scripts are released under the GNU Public License and are freely availible for use and / or modification.

__Misc__

A collection of small scripts that do various things.

__podcast__

A small perl podcast client.

__gsonginfo__

This is the GUI portion of my current song database system. It allows song artists and titles to be edited in the database. It sports another version of ObjGTK, my little GTK convienence wrapper.

__eGold__

eGold is the front-end to a content management system. It does cool template parsing and image thumbnailing.

__eMp3__

eMp3 is a little mp3 player I wrote. It's threaded to allow playlist manipulation while files are playing. It also supports logging files to a SQL database.

__geCam__
geCam is a graphical frontend for taking a picture from Video4Linux and inserting it into a database.

__eCam__

eCam is a webcam platform capable of supporting many cameras. It stores images in a MySQL database, and only uploads sufficiently changed images in order to conserve bandwidth.

__eThreads (Not included here)__

eThreads is a collection of scripts that provide multi-threaded, multi-user, database-driven forums for the World Wide Web. Currently in active development, the scripts are downloadable for use on your servers.

__camera.jpg__

camera.jpg is your standard, run-of-the-mill, no-frills web cam script. It takes a picture and prints it out.

__Distortion Cam__

distort.jpg is a variation of the camera.jpg script. It takes the snapshot and applies a random filter to it, producing weird and sometimes cool results.

__hoserAIMa IM bot__

The hoserAIMa IM bot is a basic bot that uses the AOL instant messenger servers. The bot can communicate with multiple people and store state for all of them. It currently has three 'personalities', and switches between them frequently.

podcast/Changelog
# do not edit -- automatically generated by arch changelog
# arch-tag: automatic-ChangeLog--e@ericrichardson.com--projects/podcast--main--0.1
#

2005-03-08 23:22:10 GMT	Eric Richardson (sark) <e@ericrichardson.com>	patch-1

    Summary:
      Added README File
    Revision:
      podcast--main--0.1--patch-1

    * Added README file

    new files:
     .arch-ids/Changelog.id .arch-ids/README.id Changelog README


2005-03-08 23:03:09 GMT	Eric Richardson (sark) <e@ericrichardson.com>	base-0

    Summary:
      Initial import
    Revision:
      podcast--main--0.1--base-0

    Initial import of podcast app.  Currently supports downloading audio files 
    to local machine, and (semi-untested) deleting old files to keep a certain 
    number of current files.
    

    new files:
     config podcast




eCam/camera.jpg.pl
#!/usr/bin/perl
#-----------------------------------------------------------------------------
#       ___            _   _               ___         _      _   _
#  ___ / __|__ _ _ __ (_) | |__ _  _   ___/ __| __ _ _(_)_ __| |_(_)_ _  __ _
# / -_) (__/ _` | '  \ _  | '_ \ || | / -_)__ \/ _| '_| | '_ \  _| | ' \/ _` |
# \___|\___\__,_|_|_|_(_) |_.__/\_, | \___|___/\__|_| |_| .__/\__|_|_||_\__, |
#                               |__/                    |_|             |___/
#-----------------------------------------------------------------------------
#  $Id: camera.jpg,v 1.1.1.1 2000/06/10 19:09:24 eric Exp $
#
#  eCam - a smarter webcam
#  Copyright (C) 2000 Eric Richardson
#
#     This program is free software; you can redistribute it and/or
#     modify it under the terms of the GNU General Public License
#     as published by the Free Software Foundation; either version 2
#     of the License, or (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program; if not, write to the Free Software
#     Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
#     02111-1307, USA.
#
#       For information, contact eScripting:
#           info@escripting.com
#           http://escripting.com
#
#  This script grabs the inputted host's newest picture out of the 
#  db, wraps a pre-expired expire time around it, and prints it out 
#  as an image.
#
#---------------------------------------------------------------------

use DBI;
%input = &get_vars;

my $db = DBI->connect("DBI:mysql:eCam:localhost","eCam","passwd");

my $get_ts = $db->prepare("
    select max(timestamp) from cams where host = ?
");

$get_ts->execute($input{host});

my $timestamp = $get_ts->fetchrow_array;

my $get_image = $db->prepare("
    select image from cams where host = ? and timestamp = ?
");

$get_image->execute($input{host},$timestamp);
my $img = $get_image->fetchrow_array;

print "Last-Modified: " . format_http_time(time) . "\n";
print "Expires: " . format_http_time(compute_expires_time($timestamp)) . "\n";
print "Content-Type: image/jpeg\n";
print "Content-Length: " . length($img) . "\n";
print "Pragma: no-cache\n";
print "\n";

print $img;

#----------

sub compute_expires_time {
    my $max_age = 10;
    my ($mod_time) = (@_);
    my $now = time;
    if ($mod_time < $now) { $mod_time = $now; }
    return ($mod_time - 86400);
}

#----------

sub format_http_time {
    my @time_fmt_days = ("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat");
    my @time_fmt_months = (
        "Jan", "Feb", "Mar", "Apr", "May", "Jun",
        "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
    );

    my ($time) = @_;
    my @t = gmtime($time);
    my ($sec, $min, $hour, $mday, $mon, $year, $wday) = @t;
    $year += 1900;
    $wday = $time_fmt_days[$wday];
    $mon = $time_fmt_months[$mon];
    return sprintf("%s, %02d %s %d %02d:%02d:%02d GMT",
        $wday, $mday, $mon, $year, $hour, $min, $sec);
}

#----------

sub get_vars {
    my ($info,%input);
    if ($ENV{'REQUEST_METHOD'} eq "POST") {
        read(STDIN,$info,$ENV{"CONTENT_LENGTH"});
    } else {
        $info=$ENV{QUERY_STRING};
    }

    foreach (split(/&/,$info)) {
        my ($var,$val) = split(/=/,$_,2);
        $var =~ s/\+/ /g;
        $val =~ s/\+/ /g;
        $val =~ s/%([0-9,A-F]{2})/sprintf("%c",hex($1))/ge;
        $input{$var} .= ", " if (defined($input{$var}));
        $input{$var} .= $val;
    }

    return %input;
}

#-------------#
# Change Logs #
#-------------#

# $Log: camera.jpg,v $
# Revision 1.1.1.1  2000/06/10 19:09:24  eric
# * first checkin of eCam
#

#---------------#
# End of Script #
#---------------#


gsonginfo/ObjGTK.pm
#---------------------------------------------------------------------
#  $Id: ObjGTK.pm,v 1.2 2002/03/01 07:03:16 eric Exp $
#
#  ObjGTK: GTK-Perl made simple
#  Copyright (C) 2002 Eric Richardson
#
#       This program is free software; you can redistribute it and/or
#       modify it under the terms of the GNU General Public License
#       as published by the Free Software Foundation; either version 2
#       of the License, or (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
#       02111-1307, USA.
#
#---------------------------------------------------------------------

package ObjGTK;

use Gtk;
use strict;
use vars qw(@ISA @EXPORT @EXPORT_OK);

use Exporter;
@ISA = qw(Exporter);
@EXPORT = qw();
@EXPORT_OK = qw();

#--------------------#
# Some Documentation #
#--------------------#

=head1 NAME 

ObjGTK - A GTK-Perl Wrapper

=head1 SYNOPSIS

Coming.

=head1 DESCRIPTION

This module wraps GTK to ease development.

=cut

#-------------#
# Module Core #
#-------------#

=item C<start>

    $gui = ObjGTK->start();

=cut

sub start {
    my $class = shift;

    init Gtk;

    $class = bless ( { }, $class );

    return $class;	
}

#----------

sub go {
    Gtk->main;
}

#----------

sub create_window {
    my $class = shift;
    my %args = @_;

    $class->{$args{name}} = new Gtk::Window($args{type});
    $class->{$args{name}}->set_title($args{title});

    $class->{$args{name}}->set_default_size(@{$args{d_size}}) if (
        $args{d_size}
    );

    $class->{$args{name}}->signal_connect("destroy", $args{destroy}) if (
        $args{destroy}
    );

    # now the fun begins

    $class->build($class->{$args{name}},$class->{$args{name}},$args{contents});

}

#----------

sub build {
    my ($class,$parent,$tparent,$ref) = @_;
    foreach my $child (@{$ref}) {
        my $target = "Gtk::$$child[1]";

        $tparent->{$$child[0]} = new $target(@{$$child[2]});
        $parent->add($tparent->{$$child[0]});
        $tparent->{$$child[0]}->show;

        if ($$child[3]) {
            $class->build(
                $tparent->{$$child[0]},
                ($$child[0]) ? $tparent->{$$child[0]} : $tparent,
                $$child[3]
            );
        }
    }
}

#----------

sub button_bar {
    my $class = shift;
    my $widget = shift;
    my @buttons = @_;
    
    foreach my $button (@buttons) {
        my $b = Gtk::Button->new($$button[0]);
        $b->signal_connect("clicked",$$button[1]) if ($$button[1]);
        $b->show;
        $widget->pack_start($b,0,0,0);
    }
}

#----------

sub create_radiogroup {
    my $class = shift;
    my $widget = shift;
    my @buttons = @_;
    my $m;

    foreach my $b (@buttons) {
        $widget->{$$b[0]} = Gtk::RadioButton->new(($m)?($$b[1],$$m):$$b[1]);
        $m = \$widget->{$$b[0]};
        $widget->add($widget->{$$b[0]});
        $widget->{$$b[0]}->show;
    }
}

#----------

sub menu {
    my $class = shift;
    my $widget = shift;
    my @items = @_;

    foreach my $item (@items) {
        my $i = Gtk::MenuItem->new($$item[0]);
        $i->signal_connect("activate",$$item[1],$$item[2]);
        $i->show;
        $widget->append($i);
    }
}

#----------

# $Log: ObjGTK.pm,v $
# Revision 1.2  2002/03/01 07:03:16  eric
# * GPL and CVS tags added
#

#----------

1;


distort_jpg/distort.jpg.pl
#!/usr/bin/perl
#---------------------------------------------------------------------
#  $Id$
#
#  distort.jpg - the eWorld distortion cam
#  Copyright (C) 1999 Eric Richardson
#
#     This program is free software; you can redistribute it and/or
#     modify it under the terms of the GNU General Public License
#     as published by the Free Software Foundation; either version 2
#     of the License, or (at your option) any later version.
#
#     This program is distributed in the hope that it will be useful,
#     but WITHOUT ANY WARRANTY; without even the implied warranty of
#     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#     GNU General Public License for more details.
#
#     You should have received a copy of the GNU General Public License
#     along with this program; if not, write to the Free Software
#     Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
#     02111-1307, USA.
#
#     For information on distort.jpg, contact eScripting at
#     escripting@escripting.com , or check out the distort.jpg web 
#     site at http://www.escripting.com/distort.jpg/
#
#  This script takes a picture, does some distorting, and prints out 
#  jpg data.  It should be called like a regular image (ie. <img>).
#
#---------------------------------------------------------------------

use eModule;
%input = eModule->get_vars;

%cfg = (
    away_pics_dir	=> "/usr/local/web/eric/docs/images/away-messages",
    tmp_dir			=> "/tmp/eWorld-distortcam/",
);

$time = time;

# set up some temporary files
$tmp_ppm 	= $cfg{tmp_dir} 		. $time . ".ppm";
$tmp_jpg 	= $cfg{tmp_dir} 		. $time . ".jpg";
$tmp_ppm2 	= $cfg{tmp_dir}	. "2" 	. $time . ".ppm";
$tmp_ppm3 	= $cfg{tmp_dir} . "3" 	. $time . ".ppm";
$tmp_ppm4 	= $cfg{tmp_dir} . "4" 	. $time . ".ppm";

if (&status && !$input{away}) {
    # xscreensaver isn't active
    &get_random_colors;
    &take_picture;
    &load_distorts;
    &distort;
    &convert_to_jpg;
    &upload($tmp_jpg);
    &clean_up;
} else {
    # xscreensaver is active, so there's no one here 
    &load_away_pics;
}

#-------------#
# Subroutines #
#-------------#

sub status {
    my $output = `export DISPLAY=:0; xscreensaver-command -time 2>/dev/null`;
    if ($output =~ /non-blanked/) {
        return 1;
    } else {
        return 0;
    }
}

#----------

sub take_picture {
    my $output = `streamer -o $tmp_ppm 2>&1`;
    &take_picture if ($output =~ /Device or resource busy/);
}

#----------

sub clean_up {
    `rm $tmp_ppm $tmp_jpg $tmp_ppm2 $tmp_ppm3 $tmp_ppm4 2>/dev/null`;
}

#----------

sub load_away_pics {
    my $dir = $cfg{away_pics_dir};
    opendir(DIR, "$dir");
        my @away_pics = grep{/\.jpg$/} readdir(DIR);
    closedir DIR;

    my $num_away_pics = @away_pics;
    my $n = int(rand() * $num_away_pics);
    &upload("$dir/$away_pics[$n]");
}

#----------

sub get_random_colors {
    srand(time ^ $$);
    $colors = sprintf(
        "rgb:%02x/%02x/%02x-rgb:%02x/%02x/%02x",
        int(rand()*60),
        int(rand()*60),
        int(rand()*60),
        120+int(rand()*135),
        120+int(rand()*135),
        120+int(rand()*135)
    );
    return $colors;
}

#----------

sub distort {
    srand(time ^ $$);
    $n = int(rand() * 7);
    `$distorts{$n} > $tmp_ppm4 2>/dev/null` if ($distorts{$n});
}

#----------

sub convert_to_jpg {
    `convert $tmp_ppm4 $tmp_jpg 2>/dev/null`;
}

#----------

sub upload {
    my $img = shift;
    open (IMG, "$img") || die "doh. $!\n";
        while (<IMG>) { $image_data .= $_; }
    close (IMG) || die "couldn't close $image_jpg: $!\n";

    print "Last-Modified: " . format_http_time(time) . "\n";
    print "Expires: " . format_http_time(compute_expires_time($mod_time)) . "\n";
    print "Content-Type: image/jpeg\n";
    print "Content-Length: " . length($image_data) . "\n";
    print "\n";

    print $image_data;
    
}

#----------

sub compute_expires_time {
    my ($mod_time) = (@_);
    my $now = time;
    if ($mod_time < $now) { $mod_time = $now; }
    return $mod_time + $max_age;
}

#----------

sub format_http_time {
    my @time_fmt_days = ("Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat");
    my @time_fmt_months = (
        "Jan", "Feb", "Mar", "Apr", "May", "Jun",
        "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
    );

    my ($time) = @_;
    my @t = gmtime($time);
    my ($sec, $min, $hour, $mday, $mon, $year, $wday) = @t;
    $year += 1900;
    $wday = $time_fmt_days[$wday];
    $mon = $time_fmt_months[$mon];
    return sprintf("%s, %02d %s %d %02d:%02d:%02d GMT",
        $wday, $mday, $mon, $year, $hour, $min, $sec);
}

#----------

sub load_distorts {
    my $w_h=`head -2 $tmp_ppm | tail -1`;
    ($width,$height) = split(" ",$w_h);
    %distorts = (
        0	=> "ppmtopgm $tmp_ppm | pgmedge | pgmtoppm $colors | ppmnorm",
        1	=> "ppmtopgm $tmp_ppm | pgmenhance | pgmtoppm $colors",
        2	=> "ppmtopgm $tmp_ppm | pgmoil | pgmtoppm $colors",
        3	=> "ppmrelief $tmp_ppm | ppmtopgm | pgmedge | ppmrelief | ppmtopgm | pgmedge | pnminvert | pgmtoppm $colors",
        4	=> "ppmspread 71 $tmp_ppm > $tmp_ppm2; pnmarith -add $tmp_ppm $tmp_ppm2",
        5	=> "ppmpat -g2 $width $height | pnmdepth 255 > $tmp_ppm2;pnmarith -difference  $tmp_ppm $tmp_ppm2",
        6	=> "ppmpat -anticamo $width $height | pnmdepth 255 > $tmp_ppm2; pnmarith -difference $tmp_ppm $tmp_ppm2",
        7	=> "pgmnoise $width $height | pgmedge | pgmtoppm $colors > $tmp_ppm2;pnmarith -difference $tmp_ppm $tmp_ppm2 | pnmdepth 255 | pnmsmooth",
    );
}


eGold/core.pm
#------------------------------------------------------------------------
#           .               .    .        regenerate RealGold indexes and 
# ,-,-. ,-. | , ,-. ,-. ,-. |  ,-|          new stories from the database
# | | | ,-| |<  |-' | | | | |  | | 
# ' ' ' `-^ ' ` `-' `-| `-' `' `-^ 
#                    `'                                by eric richardson
#------------------------------------------------------------------------
#  $Id: core.pm,v 1.12 2000/09/19 12:43:23 eric Exp $
#
#  makegold: generate RealGold stories and indexes
#  Copyright (C) 2000 Eric Richardson
#
#       This program is free software; you can redistribute it and/or
#       modify it under the terms of the GNU General Public License
#       as published by the Free Software Foundation; either version 2
#       of the License, or (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program; if not, write to the Free Software
#       Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  
#       02111-1307, USA.
#
#  This script looks in the RealGold stories database for new stories, 
#  and generates pages for the stories it finds.  It then also 
#  regenerates the indexes for each of the topics.
#
#------------------------------------------------------------------------

package eGold::core;

use strict;
use vars qw($db %cfg @ISA @EXPORT @EXPORT_OK @files %stored);

use Image::Magick;
use Storable;
use String::CRC32;
use DBI;
use Mail::Sendmail;
use File::Rsync;

use Exporter;
@ISA = qw(Exporter);
@EXPORT = qw(%cfg $db @files %stored);
@EXPORT_OK = qw();

#----------

=head1 NAME

eGold::core - eGold core functions

=head1 SYNOPSIS

my $result = $core->function(@args);?

=head1 DESCRIPTION

This package provides functions utilized by scripts in the eGold content
management system.

=cut

#-----------#
# Functions #
#-----------#

=item B<start>

    $core = eGold::core->start("script");

This function initializes this package, loads the configuration file 
into memory, and returns a blessed reference.

=cut

sub start {
    my $class = shift;

    # load the config file
    require "cfg.main";
    %cfg = &cfg::cfg;
    $cfg{script} = shift;

    $db = DBI->connect(
        "DBI:mysql:$cfg{db}{db}:$cfg{db}{host}",$cfg{db}{user},$cfg{db}{pass}
    );

    if (-e $cfg{cache}.$cfg{script}) {
        my $s_ref = retrieve($cfg{cache}.$cfg{script});
        %stored = %{$s_ref};
    } else {
        %stored = ();
    }

    return bless ( { }, $class );
}

#----------

=item B<save_memory>

    $core->save_memory;

Save stored memory to a file.

=cut

sub save_memory {
    my $class 	= shift;
    Storable::store(\%stored,$cfg{cache}.$cfg{script});
}

#----------

=item B<store>

    $core->store('key','value');

Add a key/value to the stored hash.

=cut

sub store {
    my $class = shift;
    $stored{$_[0]} = $_[1];
}

#----------

=item B<get_template>

    my $html = $core->get_template($fully_expanded_file);

Load a template file.  Feed get_template a fully expanded template name 
(story.locale.topic) and it'll try to load story.locale.topic.html, then 
story.locale.html, then story.html before failing.

=cut

sub get_template {
    my ($class,$template) = @_;

    $template =~ s!\.$!!;

    sub try_template {
        return unless (${$_[0]});
        if (!-e $cfg{template_dir}.${$_[0]}.".html") {
            ${$_[0]} =~ s!\.?[^\.]+$!!i;
            &try_template;
        }
    }

    &try_template(\$template);
    return unless ($template);

    open(FILE, $cfg{template_dir}.$template.".html") 
        or die "can't open template '".$cfg{template_dir}.$template.".html': $!\n";

    my $html;
    while (<FILE>) { $html .= $_; }
    return $html;
}

#----------

=item B<write_file>

    $core->write_file(
        name	=> "name",
        data	=> $data,
    );

Checksums the file and compares it to the stored value to see if the file has been 
altered before writing it.  Also compares the checksum with the checksum of $data 
to see if you're actually feeding it new data that it should upload.  

=cut

sub write_file {
    my $class 	= shift;
    my %args 	= @_;

    my $name = $class->clean_up($args{name});

    my ($dir) = $name =~ m!(.*)/[^/]+$!;

    if (-e $cfg{output_dir}.$name) {
        if ($class->different($name)) {
#			$class->email(
#				To		=> $cfg{admin_email},
#				Subject	=> "File Conflict",
#				Body	=> qq(),
#			);
            $name =~ s!\.shtml$!\.new\.shtml!;
        } else {
            return if (crc32($args{data}) == $stored{$name});
        }
    }

    warn "outfile being written: $name\n";

    $class->mkdir($dir);

    $stored{$name} = crc32($args{data});
    warn "stored crc: $stored{$name}\t$name\n";


    # this makes sure it gets rsync'ed later
    push @files, $cfg{output_dir}.$name;

    open (FILE, ">".$cfg{output_dir}.$name) 
        or die "Could not write '$name': $!\n";

    print FILE $args{data};
    close FILE;

}

#----------

=item B<email>

    $core->email(
        To		=> "whoever",
        Subject	=> "Whatever",
        Body	=> "hey",
    );

Send an email.

=cut

sub email {
    my $class = shift;

#	sendmail(
#		From	=> $cfg{admin_email},
#		@_,
#	);
}

#----------

=item B<clean_up>

    my $result = $core->clean_up($var);

Clean up a value, most likely for use as a filename.  

=cut

sub clean_up {
    my ($class) = shift;

    $_ = shift;
        s!\s!_!gi;
        s![^\w\d\./_-]!!gi;
    return $_;
}

#----------

=item B<different>

    if ($core->different("file")) {
        #do something
    } else { #whatever }

Compare the stored checksum of file with a newly computed checksum for file to 
see if someone else has edited it.

=cut

sub different {
    my ($class,$file) = @_;

    open (FILE, $cfg{output_dir}.$file);
        my $f_crc = crc32(*FILE);
    close FILE;

    ($stored{$file} == $f_crc) ? 0 : 1;
}

#----------

=item B<mkdir>

    $core->mkdir("my/sub/dir");

Create a directory and all parent directories which don't already exist.

=cut

sub mkdir {
    my $class = shift;
    my $dir = shift;

    return if (-d $cfg{output_dir}.$dir);

    my ($parent) = $dir =~ m!^(.*)/[^/]+$!;
    if (!-d $cfg{output_dir}.$parent && $parent) {
        $class->mkdir($parent);
    }

    mkdir $cfg{output_dir}.$dir,0755 or die "couldn't make $dir: $!\n";
}

#----------

=item B<thumbnail>

    $core->thumbnail(
        image	=> "img.jpg",
        thumb	=> "thumbnail.jpg",
        width	=> "max_width",
        height	=> "max_height",
    );

Take a full sized image file and generate a thumbnail out of it.

=cut

sub thumbnail {
    my $class = shift;
    my %args = @_;

    my $im = Image::Magick->new;
    $im->Read($cfg{output_dir}.$args{image});

    my ($x,$y) = $im->Get('width','height');
    
    if ($args{width} && ($args{width} < $x)) {
        my $percent = ($args{width} / $x);
        $x = sprintf("%1d",($x * $percent));
        $y = sprintf("%1d",($y * $percent));
    }

    if ($args{height} && ($args{height} < $y)) {
        my $percent = ($args{height} / $y);
        $x = sprintf("%1d",($x * $percent));
        $y = sprintf("%1d",($y * $percent));
    }

    $im->Resize(
        width	=> $x,
        height	=> $y,
    );

    if ($args{letterbox}) {
        my $foo = new Image::Magick;
        $foo->Set(
            size	=> $args{width}."x".$args{height}
        );
        $foo->ReadImage('xc:white');
        my ($off_x,$off_y) = ('0','0');
        $off_x = (($args{width}-$x)/2) if ($x < $args{width});
        $off_y = (($args{height}-$y)/2) if ($y < $args{height});

        warn "off_x: $off_x\toff_y: $off_y\n";

        $foo->Composite(
            image		=> $im,
            compose		=> 'over',
            geometry	=> "+".$off_x."+".$off_y,
        );

        $foo->Write($cfg{output_dir}.$args{thumb});
    } else {
        $im->Write($cfg{output_dir}.$args{thumb});
    }
    
    open (THUMB, $cfg{output_dir}.$args{thumb});
        my $t_crc = crc32(*THUMB);
    close THUMB;

    return if ($t_crc == $stored{$args{thumb}});
    $stored{$args{thumb}} = $t_crc;
}

#----------

=item B<upload>

    $core->upload(@files);

Upload the files found in @files.

=cut

sub upload {
    my $class = shift;
    my @files = @_;

    warn "uploading: @files\n";

    my $rsync = File::Rsync->new(
        rsh			=> "/usr/local/bin/ssh",
        perms		=> 1,
        links		=> 1,
        recursive	=> 1,
        dest		=> $cfg{rsync}{host}.":".$cfg{rsync}{dir},
        src			=> $cfg{output_dir},
    );

    $rsync->exec or die "couldn't rsync: ".$rsync->err."\n";

}

=head1 AUTHOR

Eric Richardson <eric@ericrichardson.com>

=head1 COPYRIGHT

Copyright (C) 2000 Eric Richardson

This program is licensed under the terms of the GNU General 
Public License.  

=cut

#-------------#
# Change Logs #
#-------------#

# $Log: core.pm,v $
# Revision 1.12  2000/09/19 12:43:23  eric
# * changed determination of publish state
# * added letterboxing code to core::thumbnail
#
# Revision 1.11  2000/09/17 11:05:42  eric
# * changed the way for determining published stories
# * finished more functionality in make_archives
#
# Revision 1.10  2000/09/15 12:12:24  eric
# * added random_thumb script for generating home page thumbnails
# * added thumbnail code to eGold::core
#
# Revision 1.9  2000/09/15 11:43:27  eric
# * changed todays_stories to just display last ten stories regardless
#   of day (not exactly truth in naming, but i'm lazy)
# * added make_archives script to generate archive pages
# * little fixes
#
# Revision 1.8  2000/09/13 22:46:36  eric
# * some fixes
# * made stored mem work again
#
# Revision 1.7  2000/09/11 09:38:43  eric
# * first working revision of todays_stories
# * updates to eGold::core
#
# Revision 1.6  2000/09/08 13:16:11  eric
# * moved upload function to eGold::core
#
# Revision 1.5  2000/09/08 12:34:34  eric
# * moved more functions over from make_gold to eGold::core
#
# Revision 1.4  2000/09/08 12:22:37  eric
# * added write_file, clean_up, different, and mkdir functions to package
#
# Revision 1.3  2000/09/08 11:32:01  eric
# * moved get_template into eGold::core
#
# Revision 1.2  2000/09/08 11:20:57  eric
# * it would be cool if there was a way to make ascii art out of valid
#   Perl.  Unfortunately, my ascii art doesn't compile.
# * fixed a little bug in where the package tag was placed in eGold::core
#
# Revision 1.1  2000/09/08 11:08:40  eric
# * small fixes
# * got rid of a couple warns
# * added start of core module and todays_stories script
#

#----------

1;
